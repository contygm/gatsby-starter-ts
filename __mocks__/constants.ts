const glossaryDefinitions: Array<GlossaryElements> = [
    {
        html: `<h3>Definition</h3><p>Lemon drops pastry danish halvah. Lemon drops pastry danish halvah.</p>`,
        frontmatter: {
            title: 'Air',
            letter: 'A',
            date: '3-2-2021',
            type: 'glossary',
            syllables: 'a·ban·don',
            phonetics: 'əˈbandən',
            similarWords: ['/blog/test', '/wiki/test'],
            relatedPosts: ['/blog/test1', '/wiki/test1'],
            tags: ['two', 'three']
        }
    },
    {
        html: `<h3>Definition</h3><p>Lemon drops pastry danish halvah. Lemon drops pastry danish halvah.</p>`,
        frontmatter: {
            title: 'Atom',
            letter: 'A',
            date: '3-2-2021',
            type: 'glossary',
            syllables: 'a·ban·don',
            phonetics: 'əˈbandən',
            similarWords: ['/blog/test', '/wiki/test'],
            relatedPosts: ['/blog/test1', '/wiki/test1'],
            tags: ['two', 'one']
        }
    },
    {
        html: `<h3>Definition</h3><p>Lemon drops pastry danish halvah. Lemon drops pastry danish halvah.</p>`,
        frontmatter: {
            title: 'Cloud',
            letter: 'C',
            date: '3-2-2021',
            type: 'glossary',
            syllables: 'a·ban·don',
            phonetics: 'əˈbandən',
            similarWords: ['/blog/test', '/wiki/test'],
            relatedPosts: ['/blog/test1', '/wiki/test1'],
            tags: ['two', 'one']
        }
    },
    {
        html: `<h3>Definition</h3><p>Lemon drops pastry danish halvah. Lemon drops pastry danish halvah.</p>`,
        frontmatter: {
            title: 'Moon',
            letter: 'M',
            date: '3-2-2021',
            type: 'glossary',
            syllables: 'a·ban·don',
            phonetics: 'əˈbandən',
            similarWords: ['/blog/test', '/wiki/test'],
            relatedPosts: ['/blog/test1', '/wiki/test1'],
            tags: ['two']
        }
    },
    {
        html: `<h3>Definition</h3><p>Lemon drops pastry danish halvah. Lemon drops pastry danish halvah.</p>`,
        frontmatter: {
            title: 'Window',
            letter: 'W',
            date: '3-2-2021',
            type: 'glossary',
            syllables: 'a·ban·don',
            phonetics: 'əˈbandən',
            similarWords: ['/blog/test', '/wiki/test1'],
            relatedPosts: ['/wiki/test', '/blog/test1'],
            tags: ['three']
        }
    }
];

const featuredOrRelatedPosts: Array<IndexElements> = [
    {
        excerpt:
            'Lemon drops pastry danish halvah. Chupa chups fruitcake powder...',
        fields: {
            slug: '/baking-tutorial/'
        },
        frontmatter: {
            title: 'Baking Tutorial',
            date: '3-2-2021',
            description: 'a baking tutorial',
            tags: ['two', 'three'],
            headerImage: {
                childImageSharp: {
                    gatsbyImageData: {}
                }
            }
        }
    },
    {
        excerpt:
            'Cheesecake lemon drops Cheesecake lemon drops pastry danish halvah...',
        fields: {
            slug: '/ten-best-things/'
        },
        frontmatter: {
            title: '10 Best Things',
            description: 'a list of the ten best things',
            tags: ['one', 'two', 'three'],
            date: '2-2-2021',
            headerImage: {
                childImageSharp: {
                    gatsbyImageData: {}
                }
            }
        }
    },
    {
        excerpt: 'Chupa chups fruitcake powder dessert gingerbread gummies...',
        fields: {
            slug: '/guide/'
        },
        frontmatter: {
            title: 'Guide',
            description: 'a guide',
            tags: ['one', 'two'],
            date: '5-2-2021',
            headerImage: {
                childImageSharp: {
                    gatsbyImageData: {}
                }
            }
        }
    }
];

const previous = {
    fields: {
        slug: '/post-before'
    },
    frontmatter: {
        title: 'Before - A Post About Things'
    }
};

const next = {
    fields: {
        slug: '/post-after'
    },
    frontmatter: {
        title: 'After - A Post About Things'
    }
};

const siteData: SiteMetadata = {
    title: `Starter`,
    siteUrl: `https://www.domain.com`,
    author: {
        name: `Writer B. Author`,
        description:
            'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla quam velit, vulputate eu pharetra nec, mattis ac neque. Duis vulputate commodo lectus, ac blandit elit tincidunt id. Sed rhoncus, tortor sed eleifend tristique, tortor mauris molestie elit'
    },
    businessName: 'The Company',
    phone: '999-999-9999',
    email: 'email@email.com',
    description: 'a super duper site!',
    image: '/nothing.png'
};

export {
    featuredOrRelatedPosts,
    previous,
    next,
    siteData,
    glossaryDefinitions
};
