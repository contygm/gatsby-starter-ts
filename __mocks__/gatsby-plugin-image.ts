import React from 'react';
const plugin = jest.requireActual('gatsby-plugin-image');

const mockImage = ({ imgClassName, ...props } : {imgClassName: string }) =>
    React.createElement('img', {
        ...props,
        className: imgClassName
    });

module.exports = {
    ...plugin,
    GatsbyImage: jest.fn().mockImplementation(mockImage),
    StaticImage: jest.fn().mockImplementation(mockImage)
};
