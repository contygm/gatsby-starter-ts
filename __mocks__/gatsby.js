const React = require('react');
const gatsby = jest.requireActual('gatsby');

// NOTE: This mocks the graphql() function, Link component, and StaticQuery component.
// Allow testing components that use Link or GraphQL
module.exports = {
    ...gatsby,
    graphql: jest.fn(),
    Link: jest.fn().mockImplementation(
        // these props are invalid for an `a` tag
        ({
            activeClassName,
            activeStyle,
            getProps,
            innerRef,
            partiallyActive,
            ref,
            replace,
            to,
            ...rest
        }) =>
            React.createElement('a', {
                ...rest,
                href: to
            })
    ),
    StaticQuery: jest.fn(),
    useStaticQuery: jest.fn()
};
