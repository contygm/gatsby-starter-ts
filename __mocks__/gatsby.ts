import React from 'react';
const gatsby = jest.requireActual('gatsby');

// NOTE: This mocks the graphql() function, Link component, and StaticQuery component.
// Allow testing components that use Link or GraphQL
// https://stackoverflow.com/questions/54941990/how-to-strongly-type-jest-mocks
// https://stackoverflow.com/questions/48759035/mock-dependency-in-jest-with-typescript
module.exports = {
    ...gatsby,
    graphql: jest.fn(),
    Link: jest.fn().mockImplementation(
        // these props are invalid for an `a` tag
        ({
            activeClassName,
            activeStyle,
            getProps,
            innerRef,
            partiallyActive,
            ref,
            replace,
            to,
            ...rest
        }) =>
            React.createElement('a', {
                ...rest,
                href: to
            })
    ),
    StaticQuery: jest.fn(),
    useStaticQuery: jest.fn().mockImplementation(() => ({
        site: {
            siteMetadata: {
                title: `Starter`,
                siteUrl: `https://www.domain.com`,
                author: {
                    name: `Writer B. Author`,
                    description:
                        'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla quam velit, vulputate eu pharetra nec, mattis ac neque. Duis vulputate commodo lectus, ac blandit elit tincidunt id. Sed rhoncus, tortor sed eleifend tristique, tortor mauris molestie elit'
                },
                businessName: 'The Company',
                phone: '999-999-9999',
                email: 'email@email.com',
                description: 'a super duper site!',
                image: '/nothing.png'
            }
        }
    }))
};
