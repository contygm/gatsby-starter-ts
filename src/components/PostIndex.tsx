import React, { FunctionComponent, useEffect, useState } from 'react';
import PostCard from './PostCard';

interface PostIndexProps {
    allPosts: Array<IndexElements>;
    increment: number;
}

const PostIndex: FunctionComponent<PostIndexProps> = ({
    allPosts,
    increment
}: PostIndexProps) => {
    const [displayPosts, setDisplayPosts] = useState([
        ...allPosts.slice(0, increment)
    ]);
    const [loadMore, setLoadMore] = useState(false);
    const [hasMore, setHasMore] = useState(allPosts.length > increment);

    const handleLoadMore = () => {
        setLoadMore(true);
    };

    useEffect(() => {
        if (loadMore && hasMore) {
            const postCount = displayPosts.length;
            const stillHasMore = postCount < allPosts.length;
            const nextResults = stillHasMore
                ? allPosts.slice(postCount, postCount + increment)
                : [];
            // combine old displayPosts with next batch of posts
            setDisplayPosts([...displayPosts, ...nextResults]);
            setLoadMore(false);
        }
    }, [loadMore, hasMore]);

    useEffect(() => {
        const stillHasMore = displayPosts.length < allPosts.length;
        setHasMore(stillHasMore);
    }, [displayPosts]);

    return (
        <article className="section">
            <section className="container is-max-desktop">
                {/* post cards */}
                <div className="columns is-multiline is-centered">
                    {displayPosts.map((post: IndexElements) => {
                        return (
                            <div
                                className="column is-4"
                                key={post.frontmatter.title}
                            >
                                <PostCard post={post} />
                            </div>
                        );
                    })}
                </div>
                {hasMore && (
                    <div className="columns is-mobile is-centered">
                        <button
                            className="button is-dark"
                            onClick={handleLoadMore}
                        >
                            Load more...
                        </button>
                    </div>
                )}
            </section>
        </article>
    );
};

export default PostIndex;
