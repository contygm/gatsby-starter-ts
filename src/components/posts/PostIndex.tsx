import React, { useEffect, useState } from 'react';
import { PostCard } from './PostCard';

/**
 * All props needed for the PostIndex component.
 * @property {IndexElements[]} allPosts - a list of allPosts available. This included front matter and other index elements
 * @property {number} increment - when the user clicks the 'see more posts' button, this is how many more posts should show
 * @property {PostType} type - the type of post
 * @property {void} handleFilterUpdate - a void function for when the a user interacts with the filter; `(e: any) => void`
 *
 * @see IndexElements
 * @see PostType
 * @see PostIndex
 * @category Components
 */
interface PostIndexProps {
    allPosts: IndexElements[];
    increment: number;
    type: PostType;
    handleFilterUpdate: (e: any) => void;
}

/**
 * A post index component that displays basic post info in PostCards. The component can also filter available posts based on tags.
 * Additionally, a 'see more button' that will display more posts. The initial amount of posts displayed as well as additional amount
 * of posts displayed with the button are both based on the increment passed in.
 * @param {PostIndexProps} props - includes allPosts, increment, type, handleFilterUpdate
 *
 * @category Components
 * @see PostIndexProps
 * @see PostCard
 */
export const PostIndex = (props: PostIndexProps) => {
    const [displayPosts, setDisplayPosts] = useState([
        ...props.allPosts.slice(0, props.increment)
    ]);
    const [loadMore, setLoadMore] = useState(false);
    const [hasMore, setHasMore] = useState(props.allPosts.length > props.increment);

    const handleLoadMore = () => {
        setLoadMore(true);
    };

    useEffect(() => {
        if (loadMore && hasMore) {
            const postCount = displayPosts.length;
            const stillHasMore = postCount < props.allPosts.length;
            const nextResults = stillHasMore
                ? props.allPosts.slice(postCount, postCount + props.increment)
                : [];
            // combine old displayPosts with next batch of posts
            setDisplayPosts([...displayPosts, ...nextResults]);
            setLoadMore(false);
        }
    }, [loadMore, hasMore]);

    useEffect(() => {
        setDisplayPosts([...props.allPosts.slice(0, props.increment)]);
    }, [props.allPosts]);

    useEffect(() => {
        const stillHasMore = displayPosts.length < props.allPosts.length;
        setHasMore(stillHasMore);
    }, [displayPosts]);

    return (
        <article className="section">
            <section className="container is-max-desktop">
                {/* post cards */}
                <div className="col-multi-wrapper is-centered">
                    {displayPosts.length > 0 ? (
                        displayPosts.map((post: IndexElements) => {
                            return (
                                <div
                                    data-testid={'post-card'}
                                    className="column is-4"
                                    key={post.frontmatter.title}
                                >
                                    <PostCard
                                        type={props.type}
                                        post={post}
                                        handleFilterUpdate={props.handleFilterUpdate}
                                    />
                                </div>
                            );
                        })
                    ) : (
                        <h2 className="title is-size-3">No results</h2>
                    )}
                </div>
                {hasMore && (
                    <div className="columns is-mobile is-centered">
                        <button
                            className="button is-dark"
                            onClick={handleLoadMore}
                        >
                            Load more...
                        </button>
                    </div>
                )}
            </section>
        </article>
    );
};
