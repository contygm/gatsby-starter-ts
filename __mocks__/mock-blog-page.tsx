/* istanbul ignore file */
import { BlogIndexProps } from '../src/pages/blog';
import { mockHeadProps, mockPageProps } from '../src/utils/test-props';
import { HeadProps, PageProps } from 'gatsby';
import { featuredOrRelatedPosts, siteData } from './constants';

const mockBlogs: BlogIndexProps = {
    site: siteData,
    allTags: {
        group: [
            { fieldValue: 'three', totalCount: 3 },
            { fieldValue: 'two', totalCount: 2 },
            { fieldValue: 'one', totalCount: 1 }
        ]
    },
    index: {
        totalCount: 3,
        nodes: featuredOrRelatedPosts
    },
    featured: {
        nodes: featuredOrRelatedPosts
    }
};

export const mockBlogPageData = mockPageProps(
    mockBlogs
) as PageProps<BlogIndexProps>;

export const mockBlogHeadData = mockHeadProps(
    mockBlogs
) as HeadProps<BlogIndexProps>;
