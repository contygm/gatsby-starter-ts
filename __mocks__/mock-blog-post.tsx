/* istanbul ignore file */
import { HeadProps, PageProps } from 'gatsby';
import { BlogPostProps } from '../src/templates/blog-post';
import { mockHeadProps, mockPageProps } from '../src/utils/test-props';
import { previous, next, featuredOrRelatedPosts, siteData } from './constants';

export const mockBlogPost: BlogPostProps = {
    site: {
        siteMetadata: siteData
    },
    markdownRemark: {
        id: 'slkj-dsjflks-dflkjl',
        excerpt:
            'Lemon drops pastry danish halvah. Chupa chups fruitcake powder...',
        html: `<h2>Header</h2><p>Lemon drops pastry danish halvah. Lemon drops pastry danish halvah. Lemon drops pastry danish halvah. Lemon drops pastry danish halvah.</p>`,
        tableOfContents: `<ul><li><p><a href="#liquorice-topping">Liquorice topping</a></p><ul><li><p><a href="#pastry-halvah-caramels">Pastry halvah caramels</a></p><ul> <li><a href="#wafer-oat-cake">Wafer oat cake</a></li></ul></li></ul></li></ul>`,
        frontmatter: {
            date: 'December 02, 2022',
            title: 'Baking Tutorial',
            description: 'a baking tutorial',
            related: ['/one-url/', '/two-url/', '/three-url/'],
            tags: ['one', 'two', 'three'],
            headerImage: {
                childImageSharp: {
                    gatsbyImageData: {}
                }
            }
        }
    },
    previous: previous,
    next: next,
    featured: {
        nodes: featuredOrRelatedPosts
    },
    related: {
        nodes: featuredOrRelatedPosts
    }
};

export const mockBlogPostData = mockPageProps(
    mockBlogPost
) as PageProps<BlogPostProps>;
export const mockBlogPostHeadData = mockHeadProps(
    mockBlogPost
) as HeadProps<BlogPostProps>;
