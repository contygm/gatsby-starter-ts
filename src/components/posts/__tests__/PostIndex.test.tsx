import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { PostIndex } from '../PostIndex';
import { mockBlogPageData } from '../../../../__mocks__/mock-blog-page';

describe('PostIndex', () => {
    it('default renders and load btn works correctly', () => {
        const { asFragment, getByText, getAllByTestId } = render(
            <PostIndex
                allPosts={mockBlogPageData.data.index.nodes}
                increment={2}
                type={'blog'}
                handleFilterUpdate={() => {
                    return;
                }}
            />
        );
        expect(asFragment()).toMatchSnapshot();
        expect(getAllByTestId('post-card').length).toEqual(2);

        // load more btn should load more
        fireEvent.click(getByText('Load more...'));
        expect(getAllByTestId('post-card').length).toEqual(3);
    });
});
