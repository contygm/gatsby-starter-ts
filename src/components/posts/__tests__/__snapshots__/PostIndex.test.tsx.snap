// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostIndex default renders and load btn works correctly 1`] = `
<DocumentFragment>
  <article
    class="section"
  >
    <section
      class="post-index-wrapper"
    >
      <div
        class="post-index-cards-wrapper"
      >
        <div
          class="post-index-card-wrapper"
          data-testid="post-card"
        >
          <div
            class="card"
          >
            <div
              class="card-image"
            >
              <a
                class="post-card-main-tag"
                href="?tag=two"
                id="two"
              >
                two
              </a>
              <figure
                class="post-card-image"
              >
                <img
                  alt="Placeholder image"
                  src="https://bulma.io/images/placeholders/1280x960.png"
                />
              </figure>
            </div>
            <div
              class="card-content"
            >
              <div
                class="media"
              >
                <div
                  class="media-content"
                >
                  <p
                    class="title-four"
                  >
                    Baking Tutorial
                  </p>
                  <p
                    class="subtitle-six"
                  >
                    <time
                      datetime="3-2-2021"
                    >
                      3-2-2021
                    </time>
                  </p>
                </div>
              </div>
              <div
                class="content"
              >
                <p>
                  a baking tutorial
                </p>
                <a
                  href="/blog/baking-tutorial/"
                >
                  Read more...
                </a>
              </div>
              <div
                class="card-footer"
              >
                <p
                  class="post-card-tag-title"
                >
                  Tags:
                </p>
                <div
                  class="tags"
                >
                  <a
                    class="post-card-tag"
                    href="?tag=three"
                    id="three"
                  >
                    three
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="post-index-card-wrapper"
          data-testid="post-card"
        >
          <div
            class="card"
          >
            <div
              class="card-image"
            >
              <a
                class="post-card-main-tag"
                href="?tag=one"
                id="one"
              >
                one
              </a>
              <figure
                class="post-card-image"
              >
                <img
                  alt="Placeholder image"
                  src="https://bulma.io/images/placeholders/1280x960.png"
                />
              </figure>
            </div>
            <div
              class="card-content"
            >
              <div
                class="media"
              >
                <div
                  class="media-content"
                >
                  <p
                    class="title-four"
                  >
                    10 Best Things
                  </p>
                  <p
                    class="subtitle-six"
                  >
                    <time
                      datetime="2-2-2021"
                    >
                      2-2-2021
                    </time>
                  </p>
                </div>
              </div>
              <div
                class="content"
              >
                <p>
                  a list of the ten best things
                </p>
                <a
                  href="/blog/ten-best-things/"
                >
                  Read more...
                </a>
              </div>
              <div
                class="card-footer"
              >
                <p
                  class="post-card-tag-title"
                >
                  Tags:
                </p>
                <div
                  class="tags"
                >
                  <a
                    class="post-card-tag"
                    href="?tag=two"
                    id="two"
                  >
                    two
                  </a>
                  <a
                    class="post-card-tag"
                    href="?tag=three"
                    id="three"
                  >
                    three
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="post-index-load-more-wrapper"
      >
        <button
          class="post-index-load-more"
        >
          Load more...
        </button>
      </div>
    </section>
  </article>
</DocumentFragment>
`;
